File: emulators/python/linux-emulator-example.py
#   Copyright 2016 Robert Elder Software Inc.
#   
#   Licensed under the Apache License, Version 2.0 (the "License"); you may not 
#   use this file except in compliance with the License.  You may obtain a copy 
#   of the License at
#   
#       http://www.apache.org/licenses/LICENSE-2.0
#   
#   Unless required by applicable law or agreed to in writing, software 
#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 
#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the 
#   License for the specific language governing permissions and limitations 
#   under the License.
import sys
import select
import time
import sys, tty, termios
from op_cpu_package.python_l0_module import OpCPUData
from op_cpu_package.op_cpu_module import OpCPU

#  A Linux interface to using the python implementation of the One Page CPU emulator

def main_loop():
    loader_data = OpCPUData()
    op_cpu = OpCPU(loader_data)
    fd = sys.stdin.fileno()
    old_settings = termios.tcgetattr(fd)
    tty.setraw(sys.stdin.fileno())
    input_buffer = []  #  Characters to be sent to emulator
    while not op_cpu.is_halted():
        in_chrs = select.select([sys.stdin], [], [], 0.0001)[0]
        if not in_chrs:
            for x in range(0, 10000):
                r = op_cpu.vm_getc()
                if 'chr' in r:
                    sys.stdout.write(chr(r['chr']))
                    if r['chr'] == 10:
                        sys.stdout.write('\r')
                    sys.stdout.flush()
                if len(input_buffer):
                    inchr = input_buffer.pop()
                    if op_cpu.vm_putc(inchr):  #  Not able to input chr
                        input_buffer = [inchr] + input_buffer
                
                op_cpu.step()
        else:
            dobreak = False
            for file in in_chrs:
                c = file.read(1)
                input_buffer = input_buffer + [ord(c)]
                if ord(c) == 3:
                    dobreak = True
            if dobreak:
                break
    termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)

main_loop()

File: emulators/c/linux-emulator-example.c
/*
    Copyright 2016 Robert Elder Software Inc.
    
    Licensed under the Apache License, Version 2.0 (the "License"); you may not 
    use this file except in compliance with the License.  You may obtain a copy 
    of the License at
    
        http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software 
    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 
    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the 
    License for the specific language governing permissions and limitations 
    under the License.
*/

/*
  A linux compatible interface for running .l0.c files in the op-cpu emulator.  Takes care
of non-standard terminal input, and output.

*/

#include "op-cpu.h"
#include <termios.h>
#include <unistd.h>
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>

extern unsigned char data[][5];

struct termios * terminal_setup(void);

struct termios * terminal_setup(void){
	/*  Change settings to let input be completely processed by the emulator. */
	struct termios * original = malloc(sizeof(struct termios));
        struct termios stdio;
	tcgetattr(STDOUT_FILENO, original);
 
        memcpy(&stdio, original, sizeof(struct termios));
        stdio.c_iflag = 0;
        stdio.c_cflag = 0;
        stdio.c_lflag = 0;
        stdio.c_cc[VMIN]=1;
        stdio.c_cc[VTIME]=0;
        tcsetattr(STDOUT_FILENO,TCSANOW,&stdio);
        tcsetattr(STDOUT_FILENO,TCSAFLUSH,&stdio);
        fcntl(STDIN_FILENO, F_SETFL, O_NONBLOCK);
        return original;
}

struct ll_item{
	struct ll_item * next;
	unsigned int c;
	unsigned int pad;
};

int main(void){
	struct virtual_machine * vm;
	unsigned int output;
        unsigned char c;
	struct termios * original = terminal_setup();
	struct ll_item * first_item = (struct ll_item *)0;
	struct ll_item * last_item = (struct ll_item *)0;

	vm = vm_create(data);
	printf("Kernel image has been loaded. All input is now being handled by the emulator (including Ctrl+c).\n");
	printf("Press 'q' to quit.\n");

	while(!is_halted(vm)){
		if(vm_getc(vm, &output)){
			putchar((int)output);
			fflush(stdout);
		}

		if(read(STDIN_FILENO, &c, 1) > 0){
			/*  Add item to the end of the list */
			struct ll_item * new_item = malloc(sizeof(struct ll_item));
			new_item->c = c;
			new_item->next = (struct ll_item *)0;
			if(first_item){
				last_item->next = new_item;
			}else{
				first_item = new_item;
			}
			last_item = new_item;
		}

		if(first_item){
			if(!vm_putc(vm, first_item->c)){
				/*  Take first item off of linked list */
				struct ll_item * tmp_item = first_item;
				first_item = first_item->next;
				free(tmp_item);
				if(!first_item){
					last_item = (struct ll_item *)0;
				}
			}
		}
		step(vm);
	}

	vm_destroy(vm);
        tcsetattr(STDOUT_FILENO,TCSANOW, original);
        tcsetattr(STDOUT_FILENO,TCSAFLUSH, original);
        free(original);
	return 0;
}
